### DISCUSSION TITLE
Do I have a brick? I'm unable to upload sketches after using this library!

### DISCUSSION THREAD

**Discussion by sailorsail** [khoih-prog/SAMD_TimerInterrupt#21] [Q&A] [answered]
I've tried to use this library (1.7.0) on an Arduino MKR WiFi 1010, and after uploading a simple sketch that uses this library, to blink a led, I'm unable to upload any new sketch!

I'm getting the following error when uploading:
"No device found on cu.usbmodem14301
An error occurred while uploading the sketch"

This is the uploaded code:
```
#include "SAMDTimerInterrupt.h"
SAMDTimer ITimer0(TIMER_TC3);
const byte led = 3;

void TimerHandler0(void) {
  digitalWrite(led, !digitalRead(led));
  Serial.println("Time's up!");
  }

#define TIMER0_INTERVAL_MS        1000      // 1s = 1000ms

void setup() {
  if (ITimer0.attachInterruptInterval(TIMER0_INTERVAL_MS * 1000, TimerHandler0))
    Serial.println("Starting  ITimer0 OK, millis() = " + String(millis()));
  else
    Serial.println("Can't set ITimer0. Select another freq. or timer");
  }  

void loop() {
}
```

I do see "Time's up!" being printed at a very high speed!
It looks like that the timer interval is not set properly, and so, while trying to upload a new sketch, the serial port doesn't have a chance to be used! Does it makes sense?

Do you know if there is any thing wrong with this code?
Do you have any clue that could help on solving this deadlock, or do I have a brick?
Thank you.

**Comment by khoih-prog** [comment]
> All Arduino boards have a built-in bootloader which allows flashing the board via USB. In case a sketch locks up the processor and the board is not reachable anymore via USB it is possible to enter bootloader mode by double-tapping the reset button right after power
> up.

**Reply by sailorsail** [reply]
Thank you, you are very right. I don't have a brick any more.
Now I was able to trace a bit further the problem I've, when setting the Timer using **attachInterruptInterval()**, with this code:
```cpp
#include "SAMDTimerInterrupt.h"
SAMDTimer ITimer0(TIMER_TC3);

int checkTimer = 0;

void TimerHandler0(void) {
  // Doing something here inside ISR
  checkTimer = checkTimer + 1;
  }

void setup() {
  Serial.begin(9600);
  while (!Serial);
  if (ITimer0.attachInterruptInterval(80000, TimerHandler0))
    Serial.println("Starting  ITimer0 OK, millis() = " + String(millis()));
  else
    Serial.println("Can't set ITimer0. Select another freq. or timer");
  }  

void loop() {
  Serial.println("Timeout's # = " + String(checkTimer));
  delay(1000);
  }
```

If I set interval to 20000, the timeout occurs 50 times per second, as expected, but if I set interval to 20001 the timeout steps down to 12 times a second!
Then, when I increase the interval up to 80000 the frequency goes progressively down to 3 times per second, however if I set interval to 80001 the timeout steps up to 3200 times a second!

Do you know if I'm doing any thing wrong?
Once again, thank you for your time and prompt reply.

**Comment by khoih-prog** [comment]
There is a bug of the v1.7.0 which creates an issue for TC3 timer.

You have to change and use **TCC**, then everything will be OK now with v1.7.0 or go back to v1.6.0

```
SAMDTimer ITimer0(TIMER_TCC);
```

Remember that the maximum time you can use directly on SAMD TC3/TCC Timer is ~1398.11ms = 1.39811s.

If you use time larger than 1398.11ms, the wrap-around will happen and you'll get much smaller time.

**Reply by sailorsail** [reply]
With TCC it works fine; and yes, it wrap-around with a time larger than 1398.11ms

**Comment by khoih-prog** [comment]
If you'd like to have much longer time, check the [**ISR_16_Timers_Array_Complex**](https://github.com/khoih-prog/SAMD_TimerInterrupt/tree/main/examples/ISR_16_Timers_Array_Complex) example.

This is the main reason this library is written for.

**Comment by stripwax** [comment]
Rui - Any different behaviour when you use another version? e.g 1.6.0
Also note that it is inadvisable to use serial from inside an ISR.

**Reply by khoih-prog** [reply]
It's somewhat OK to use Serial in ISR for SAMD, certainly **for debugging purpose** only, not for normal usage. v1.6.0 is still OK to use.

I'll release new version v1.8.0 soon to fix the bug of v1.7.0

**Reply by sailorsail** [reply]
It was for debugging, and somehow it got clogged with the use of serial inside the ISR!
I haven't tried the previous release...

**Comment by khoih-prog** [comment]
@stripwax 

I'll scrap your update in v1.7.0. Can you help debug to see what's wrong with the update in v1.7.0 causing issue in TC3 as described in this Discussion.

**Reply by sailorsail** [reply]
If you guide me on what you need I can try...

**Comment by stripwax** [comment]
My changes were to illustrate potential fix for the bug report that I made
- they certainly were not fully tested for other cases.  I assumed you
would be tesying before releasing. I can try and help but I only have
access to one SAMD21 device.

**Reply by khoih-prog** [reply]
> I assumed you would be testing before releasing. 

Certainly it's my sole responsibility, and I messed this up this time because just tested for some cases (with short time 20ms).
Good lesson to learn all life.

**Comment by khoih-prog** [comment]
Anyway, in next few releases, I'll try to add more Timers as possible, hopefully time permitted, into the library, such as

1. TCC0, TCC1, TCC2, TC0, TC1, TC2, TC3 for SAMD51
2. TCC0, TCC1, TCC2, TC3, TC4, TC5 for SAMD21

**Comment by khoih-prog** [comment, marked as answer]
@sailorsail 

Just published the new [**SAMD_TimerInterrupt** releases v1.8.0](https://github.com/khoih-prog/SAMD_TimerInterrupt/releases/tag/v1.8.0).

Your Contribution is noted in [**Contributions and Thanks**](https://github.com/khoih-prog/SAMD_TimerInterrupt#contributions-and-thanks)

---

### Releases v1.8.0

1. Fix bug introduced in v1.7.0 to SAMD21 TC3. Check [Do I have a brick? I'm unable to upload sketches after using this library! #21](https://github.com/khoih-prog/SAMD_TimerInterrupt/discussions/21)

**Reply by sailorsail** [reply]
Perfect! I've tested and it is working.
Thank you once again for you prompt reaction.
It's impressive how quick you have fixed this.

**Comment by khoih-prog** [comment]
Hi @sailorsail 

Please help test the new [**SAMD_TimerInterrupt** releases v1.9.0](https://github.com/khoih-prog/SAMD_TimerInterrupt/releases/tag/v1.9.0), which added 4 more Timers (**TC4, TC5, TCC1 and TCC2**) for SAMD21

I'm sure you'll find new bug, if exists, with your thorough tests.

Have a look at new example [**SAMD21_MultiTimers**](https://github.com/khoih-prog/SAMD_TimerInterrupt/tree/main/examples/SAMD21_MultiTimers) for a demo how to use all 6 SAMD21 timers simultaneously.

---

### Releases v1.9.0

1. Add TC4, TC5, TCC1 and TCC2 Timers to SAMD21
2. Add example [**SAMD21_MultiTimers**](https://github.com/khoih-prog/SAMD_TimerInterrupt/tree/main/examples/SAMD21_MultiTimers) to demo how to use all 6 SAMD21 timers simultaneously.
3. Add functions `attachInterruptInterval_MS()` and `setInterval_MS()`
4. Rewrite examples to take advantage of new functions and timers

**Comment by stripwax** [comment]
See #17 but I think I found the bug in 1.7.0 (although oddly it looks like you already fixed it for SAMD51 but not SAMD21, but I don't have SAMD51 so can't confirm that).  I fixed the bug in my fork and provided my test results in #17 so you can compare.
